server/04-python-dev-base/install.sh

```diff
@@ -15,3 +15,10 @@ apt-get -qq -y --no-install-recommends install \
 rm -rf /var/lib/apt/lists/*
 
 dpkg --list
+
+python3 -m venv /opt/build-venv
+
+/opt/build-venv/bin/python -m pip install -q --no-cache-dir --upgrade pip
```

Comment:
if you source the activate file (or change the PATH before running this script), you don't need to qualify these - i have to believe activate does some other setup than that too, right?

also, while we're at it, is there a reason not to make these just one single install command?

Reply:
So, I can source the activate file here, but it will only exist for the duration of the script, it won't persist into the docker layers. In this context, since I'm the creator of the venv, I'm okay being explicit about it since we are the creators and have the specific knowledge of the venv.

I think there is good reason to do pip stand-alone first, as it may have effects on the later install commands. I can go ahead and combine the latter installs.

Reply:
I also lied - it *does* also make sense to have setuptools upgraded standalone in some cases. In the python-dev-base case, it doesn't actually matter, but in the server-base case it *does* - setuptools may play a role during the installation of requirements (some packages may actually be built from source, in which case setuptools is a dependency).

"Newer" version of the `venv` module have this option:

```
--upgrade-deps        Upgrade core dependencies: pip setuptools to the latest version in PyPI
```

but that is not available from the `python3-venv` we install.